README.md
# OME Data Model

[![Build Status](https://github.com/ome/ome-model/workflows/Maven/badge.svg)](https://github.com/ome/ome-model/actions)
[![Build Status](https://github.com/ome/ome-model/workflows/Tox/badge.svg)](https://github.com/ome/ome-model/actions)
[![Maven Central](https://img.shields.io/maven-central/v/org.openmicroscopy/ome-xml.svg)](http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22org.openmicroscopy%22%20AND%20a%3A%22ome-xml%22)
[![Javadocs](http://javadoc.io/badge/org.openmicroscopy/ome-xml.svg)](http://javadoc.io/doc/org.openmicroscopy/ome-xml) 
[![PyPI](https://badge.fury.io/py/ome-model.svg)](https://pypi.org/project/ome-model/)

OME data model specification, code generator and implementation.

Links
-----

- [Documentation](https://docs.openmicroscopy.org/latest/ome-model/)
- [Java API reference](http://javadoc.io/doc/org.openmicroscopy/ome-xml/)
- [OME-TIFF sample images](https://downloads.openmicroscopy.org/images/OME-TIFF/)
- [OME-XML sample images](https://downloads.openmicroscopy.org/images/OME-XML/)

More information
----------------

For more information, see the [OME Model documentation](https://docs.openmicroscopy.org/latest/ome-model/).

Pull request testing
--------------------

We welcome pull requests from anyone, but ask that you please verify the
following before submitting a pull request:

 * verify that the branch merges cleanly into ```master```
 * verify that the branch compiles using Maven
 * verify that the branch does not use syntax or API specific to Java 1.8+
 * internal developers only: [run the data
   tests](https://docs.openmicroscopy.org/latest/bio-formats/developers/commit-testing.html)
   against directories corresponding to the affected format(s)
 * make sure that your commits contain the correct authorship information and,
   if necessary, a signed-off-by line
 * make sure that the commit messages or pull request comment contains
   sufficient information for the reviewer(s) to understand what problem was
   fixed and how to test it


NEWS.md
## 6.3.1 (2022-05-19)

### Java

*  Bump logback dependencies to `ch.qos.logback:logback-core:1.2.9` and `ch.qos.logback:logback-classic:1.2.9` ([#160](https://github.com/ome/ome-model/pull/160))
*  Bump ome-common dependency to `org.openmicroscopy:ome-common:6.0.9`

### Documentation

* Update link to TIFF standard ([#157](https://github.com/ome/ome-model/pull/157))
* Add XSLT transformations page ([#154](https://github.com/ome/ome-model/pull/154))
* Add reference to javadoc.io ([#153](https://github.com/ome/ome-model/pull/153))

### Infrastructure

* Add JDK17 to the GitHub workflows testing matrix ([#159](https://github.com/ome/ome-model/pull/159))

## 6.3.0 (2021-10-08)

### Python

* ome_model: add support for setting physical sizes and units [#151](https://github.com/ome/ome-model/pull/151)

### Documentation

* Exclude external link failing the linkcheck [#150](https://github.com/ome/ome-model/pull/150)

## 6.2.3 (2021-06-23)

### Specification

* 2008-09 to 2009-09 XSLT transforms: Fix loss of CustomAttributes [#145](https://github.com/ome/ome-model/pull/145)

### Java

* MetadataConverter: Fix dropping of WellAnnotationRefs during conversion [#131](https://github.com/ome/ome-model/pull/131)

### Python

* ome_model: do not use mutable default arguments [#149](https://github.com/ome/ome-model/pull/149)

### Infrastructure

* Dependencies: bump ome-common to 6.0.7 and logback-classic/logback-core to 1.2.0
* Remove legacy Appveyor configuration [#148](https://github.com/ome/ome-model/pull/148)
* Exclude broken links to www.fei.com [#141](https://github.com/ome/ome-model/pull/141)

## 6.2.2 (2020-12-10)

### Java

* OMEModelImpl: do not permit the addition of null references [#125](https://github.com/ome/ome-model/pull/125)

### Documentation

* Fix external reference to XML validator [#137](https://github.com/ome/ome-model/pull/137)

### Infrastructure

* Switch to GitHub workflows
* Python 3.9 support: upgrade to Genshi 0.7.5 and consume upstream package [#127](https://github.com/ome/ome-model/pull/127) [#138](https://github.com/ome/ome-model/pull/138)

## 6.2.1 (2020-09-17)

### Python

* Fix PyPI deployment [#124](https://github.com/ome/ome-model/pull/124)

## 6.2.0 (2020-09-16)

### Python

* Allow version introspection [#123](https://github.com/ome/ome-model/pull/123)
* Add support for population of plane metadata [#122](https://github.com/ome/ome-model/pull/122)
* Ensure stdout works in Jupyter Notebooks [#121](https://github.com/ome/ome-model/pull/121)
* Make channel names and color optional [#120](https://github.com/ome/ome-model/pull/120)

## 6.1.2 (2020-08-12)

### Documentation

* Fix links to external resources [#116](https://github.com/ome/ome-model/pull/116) [#118](https://github.com/ome/ome-model/pull/118)
* Add HCS companion fileset to the OME-TIFF sample data page [#117](https://github.com/ome/ome-model/pull/117)

### Python

* Fix default z,c,t values of add_tiffdata() [#115](https://github.com/ome/ome-model/pull/115)
* Automatically deploy to PyPI on tags [#119](https://github.com/ome/ome-model/pull/119)

## 6.1.1 (2020-06-04)

* Fix various documentation hyperlinks [#112](https://github.com/ome/ome-model/pull/112) [#113](https://github.com/ome/ome-model/pull/113) [#114](https://github.com/ome/ome-model/pull/114)

## 6.1.0 (2020-04-07)

### Python 3 support

* Update code-generation for building on Python 3.6, 3.7 [#65](https://github.com/ome/ome-model/pull/65)
* Add support for Python 3.8 and make code-generation Python 3 only [#105](https://github.com/ome/ome-model/pull/105)
* Uncap Sphinx version [#109](https://github.com/ome/ome-model/pull/109)

### OME-XML

* Add getters and setters for OME@Creator attribute for ome.xml.metadata interfaces and implementations [#108](https://github.com/ome/ome-model/pull/108)

### C++

* Remove unmaintained C++ OME-XML implementation [#111](https://github.com/ome/ome-model/pull/111)

## 6.0.1 (2019-05-09)

### Java

* Fix cyclic dependency between specification and ome-xml modules [#101](https://github.com/ome/ome-model/pull/101)

## 6.0.0 (2019-02-13)


### OME-TIFF

* Add sub-resolution support to the OME-TIFF specification [#98](https://github.com/ome/ome-model/pull/98)
* Add references to public OME-TIFF sub-resolution sample datasets [#99](https://github.com/ome/ome-model/pull/99)
* Documentation and linkchecker improvements [#70](https://github.com/ome/ome-model/pull/70) [#76](https://github.com/ome/ome-model/pull/76) [#83](https://github.com/ome/ome-model/pull/83) [#94](https://github.com/ome/ome-model/pull/94)

### Java

* Require JDK 8 as the minimum version [90](https://github.com/ome/ome-model/pull/90) [95](https://github.com/ome/ome-model/pull/95)
* MetadaConverter: add API for copying Channel metadata [#88](https://github.com/ome/ome-model/pull/88)
* Add createRotationTransform static method to AffineTransform [#77](https://github.com/ome/ome-model/pull/77)
* Build system: update Maven plugin versions [#74](https://github.com/ome/ome-model/pull/74)

### C++

* Aggregate generated C++ source files for model and enums [#79](https://github.com/ome/ome-model/pull/79)
* Require C++ 14 support [#76](https://github.com/ome/ome-model/pull/75)
* Add createRotationTransform static method to AffineTransform [#77](https://github.com/ome/ome-model/pull/77)

### Python

* Add experimental OME-XML generation Python API [#91](https://github.com/ome/ome-model/pull/91) [#92](https://github.com/ome/ome-model/pull/92) [#96](https://github.com/ome/ome-model/pull/96) [#97](https://github.com/ome/ome-model/pull/97)


5.6.3 (2018-03-13)
------------------

* Build system: update Maven plugin versions and unify property names [#53](https://github.com/ome/ome-model/pull/53) [#69](https://github.com/ome/ome-model/pull/69)

5.6.2 (2018-01-25)
------------------

* Report numerical value domain errors more clearly
  [#68](https://github.com/ome/ome-model/pull/68)

5.6.1 (2018-01-17)
------------------

Documentation updates:


ome_model/experimental.py
#!/usr/bin/env python
# Generate companion files

import re
import sys
import uuid
import xml.etree.ElementTree as ET
from . import __version__

OME_ATTRIBUTES = {
    'Creator': "ome_model %s" % __version__,
    'UUID': "urn:uuid:%s" % uuid.uuid4(),
    'xmlns': 'http://www.openmicroscopy.org/Schemas/OME/2016-06',
    'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
    'xsi:schemaLocation': 'http://www.openmicroscopy.org/Schemas/OME/2016-06 \
http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd',
}

TIFF_PARSER = "^.*?"  # Ignore prefix
TIFF_PARSER += "(?:[-_]+|"  # Skip separators
TIFF_PARSER += "[Cc]_?(?P<channel>[^-_]+)|"  # Channel section
TIFF_PARSER += "[TtHhRr]_?(?P<time>[^-_]+)|"  # "Time" section
TIFF_PARSER += "[Zz]_?(?P<slice>[^-_]+)"  # Z-slices
TIFF_PARSER += ")+"  # As many of those three as needed
TIFF_PARSER += ".*?[.].*?"  # Ignore the rest, but don't slurp the file ending
TIFF_PARSER = re.compile(TIFF_PARSER)


class Channel(object):

    ID = 0

    def __init__(self,
                 image,
                 name=None,
                 color=None,
                 samplesPerPixel=1,
                 ):
        self.data = {
            'ID': 'Channel:%s' % self.ID,
            'SamplesPerPixel': str(samplesPerPixel),
        }
        if name:
            self.data["Name"] = name
        if color:
            self.data["Color"] = str(color)
        Channel.ID += 1


class Plane(object):

    ALLOWED_KEYS = (
        'DeltaT', 'DeltaTUnit', 'ExposureTime', 'ExposureTimeUnit',
        'PositionX', 'PositionXUnit', 'PositionY', 'PositionYUnit',
        'PositionZ', 'PositionZUnit')

    def __init__(self, TheC=0, TheZ=0, TheT=0, options={}):
        self.data = {
            'TheC': str(TheC),
            'TheZ': str(TheZ),
            'TheT': str(TheT),
        }
        if options:
            for key, value in options.items():
                if key in self.ALLOWED_KEYS:
                    self.data[key] = value


class UUID(object):
    def __init__(self,
                 filename=None
                 ):
        self.data = {"FileName": filename}
        self.value = "urn:uuid:%s" % str(uuid.uuid4())


class TiffData(object):
    def __init__(self,
                 firstC=0,
                 firstT=0,
                 firstZ=0,
                 ifd=None,
                 planeCount=None,
                 uuid=None
                 ):
        self.data = {
            "FirstC": str(firstC),
            "FirstT": str(firstT),
            "FirstZ": str(firstZ)
        }
        self.uuid = uuid
        if ifd:
            self.data["IFD"] = str(ifd)
        if planeCount:
            self.data["PlaneCount"] = str(planeCount)


class Image(object):

    ID = 0

    def __init__(self,
                 name,
                 sizeX, sizeY, sizeZ, sizeC, sizeT,
                 physSizeX=None, physSizeY=None, physSizeZ=None,
                 physSizeXUnit=None, physSizeYUnit=None, physSizeZUnit=None,
                 tiffs=None,
                 order="XYZTC",
                 type="uint16",
                 ):
        self.data = {
            'Image': {'ID': 'Image:%s' % self.ID, 'Name': name},
            'Pixels': {
                'ID': 'Pixels:%s:%s' % (self.ID, self.ID),
                'DimensionOrder': order,
                'Type': type,
                'SizeX': str(sizeX),
                'SizeY': str(sizeY),
                'SizeZ': str(sizeZ),
                'SizeT': str(sizeT),
                'SizeC': str(sizeC),
            },
            'Channels': [],
            'TIFFs': [],
            'Planes': [],
        }
        Image.ID += 1
        if tiffs:
            for tiff in tiffs:
                self.add_tiff(tiff)
        if physSizeX:
            self.data['Pixels']['PhysicalSizeX'] = str(physSizeX)
            if physSizeXUnit:
                self.data['Pixels']['PhysicalSizeXUnit'] = str(physSizeXUnit)
        if physSizeY:
            self.data['Pixels']['PhysicalSizeY'] = str(physSizeY)
            if physSizeYUnit:
                self.data['Pixels']['PhysicalSizeYUnit'] = str(physSizeYUnit)
        if physSizeZ:
            self.data['Pixels']['PhysicalSizeZ'] = str(physSizeZ)
            if physSizeZUnit:
                self.data['Pixels']['PhysicalSizeZUnit'] = str(physSizeZUnit)

    def add_channel(self, name=None, color=None, samplesPerPixel=1):
        self.data["Channels"].append(
            Channel(
                self, name=name, color=color, samplesPerPixel=samplesPerPixel
            ))

    def add_tiff(self, filename, c=0, t=0, z=0, ifd=None, planeCount=None):

        if c is None and t is None and z is None:
            # If no mapping specified, assume single plane TIFF which name
            # contain the z,c,t indices
            c, t, z = parse_tiff(filename)
        self.data["TIFFs"].append(TiffData(
            firstC=c,
            firstT=t,
            firstZ=z,
            ifd=ifd,
            planeCount=planeCount,
            uuid=UUID(filename)))

    def add_plane(self, c=0, t=0, z=0, options={}):
        assert c >= 0 and c < int(self.data['Pixels']['SizeC'])
        assert z >= 0 and z < int(self.data['Pixels']['SizeZ'])
        assert t >= 0 and t < int(self.data['Pixels']['SizeT'])
        self.data["Planes"].append(Plane(
            TheC=c, TheT=t, TheZ=z, options=options))

    def validate(self):
        sizeC = int(self.data["Pixels"]["SizeC"])
        assert (len(self.data["Channels"]) <= sizeC), str(self.data)
        channel_samples = sum([int(x.data['SamplesPerPixel'])
                              for x in self.data["Channels"]])
        assert channel_samples <= sizeC, str(self.data)
        return self.data


class Plate(object):

    ID = 0

    def __init__(self, name, rows=None, columns=None):
        self.data = {
            'Plate': {'ID': 'Plate:%s' % self.ID, 'Name': name},
            'Wells': [],
        }
        Plate.ID += 1
        if rows:
            self.data['Plate']['Rows'] = str(rows)
        if columns:
            self.data['Plate']['Columns'] = str(columns)

    def add_well(self, row, column):
        well = Well(self, row, column)
        self.data["Wells"].append(well)
        return well


class Well(object):

    ID = 0

    def __init__(self, plate, row, column):
        self.data = {
            'Well': {
                'ID': 'Well:%s' % self.ID,
                'Row': '%s' % row,
                'Column': '%s' % column
            },
            'WellSamples': [],
        }
        Well.ID += 1

    def add_wellsample(self, index, image):
        wellsample = WellSample(self, index, image)
        self.data["WellSamples"].append(wellsample)
        return wellsample


class WellSample(object):

    ID = 0

    def __init__(self, well, index, image):
        self.data = {
            'WellSample': {
                'ID': 'WellSample:%s' % self.ID,
                'Index': '%s' % index,
            },
            'Image': image,
        }
        WellSample.ID += 1


def parse_tiff(tiff):
    """
    Extracts (c, t, z) from a tiff filename.
    """
    m = TIFF_PARSER.match(tiff)
    return (m.group("channel"), m.group("time"), m.group("slice"))


def create_companion(plates=None, images=None, out=None):
    """
    Create a companion OME-XML for a given experiment.
    Assumes 2D TIFFs
    """
    root = ET.Element("OME", attrib=OME_ATTRIBUTES)
    if not plates:
        plates = []
    if not images:
        images = []

    for plate in plates:
        p = ET.SubElement(root, "Plate", attrib=plate.data['Plate'])
        for well in plate.data["Wells"]:
            w = ET.SubElement(p, "Well", attrib=well.data["Well"])
            for wellsample in well.data["WellSamples"]:
                ws = ET.SubElement(w, "WellSample",
                                   attrib=wellsample.data['WellSample'])
                image = wellsample.data['Image']
                images.append(image)
                ET.SubElement(ws, "ImageRef", attrib={
                    "ID": image.data['Image']["ID"]})

    for img in images:
        i = img.validate()
        image = ET.SubElement(root, "Image", attrib=i["Image"])
        pixels = ET.SubElement(image, "Pixels", attrib=i["Pixels"])
        for channel in i["Channels"]:
            c = channel.data  # TODO: validation?
            ET.SubElement(pixels, "Channel", attrib=c)

        for tiff in i["TIFFs"]:
            tiffdata = ET.SubElement(pixels, "TiffData", attrib=tiff.data)
            if tiff.uuid:
                ET.SubElement(
                    tiffdata, "UUID", tiff.uuid.data).text = tiff.uuid.value

        for plane in i["Planes"]:
            ET.SubElement(pixels, "Plane", attrib=plane.data)

    # https://stackoverflow.com/a/48671499/56887
    kwargs = dict(encoding="UTF-8")
    kwargs["xml_declaration"] = True
    if not out:
        try:
            out = sys.stdout.buffer
        except AttributeError:
            # ipykernel.iostream.OutStream (used by Jupyter Notebook) does not
            # have a .buffer attribute
            out = sys.stdout
    ET.ElementTree(root).write(out, **kwargs)


def fake_image(basename="test", sizeX=64, sizeY=64, sizeZ=1, sizeC=3, sizeT=1):
    tiffs = ["%s_z%s_c%s_t%s.tiff" % (basename, z, c, t)
             for z in range(sizeZ) for c in range(sizeC)
             for t in range(sizeT)]
    image = Image("test", sizeX, sizeY, sizeZ, sizeC, sizeT, tiffs)
    image.add_channel("red", 0)
    image.add_channel("green", 0)
    image.add_channel("blue", 0)
    return image


def fake_plate(rows=2, columns=2, fields=1):

    plate = Plate("test", rows, columns)
    for row in range(rows):
        for column in range(columns):
            well = plate.add_well(row, column)
            for field in range(fields):
                basename = "Well%s%s" % (chr(row + 65), column)
                image = fake_image(basename=basename)
                well.add_wellsample(field, image)
    return plate


if __name__ == "__main__":
    plates = [fake_plate()]
    create_companion(plates=plates)
